File: Python3/2300. Successful Pairs of Spells and Potions.py
class Solution:
    def successfulPairs(self, spells: List[int], potions: List[int], success: int) -> List[int]:
        potions.sort(reverse = True)
        ans = []

        def find(s: int) -> int:
            l, r = 0, len(potions) - 1
            while l <= r:
                m = (r - l) // 2 + l
                # m cannot be included
                if potions[m] * s < success:
                    r = m - 1
                else:
                    l = m + 1
            return l


        for s in spells:
            ans.append(find(s))
            
        return ans

File: Java/2300. Successful Pairs of Spells and Potions.java
class Solution {
    public int[] successfulPairs(int[] spells, int[] potions, long success) {
        Arrays.sort(potions);
        int p = potions.length;

        int i = 0, max = potions[p-1];
        int[] ans = new int[spells.length];
        for (int s: spells) {
            long min = (long) Math.ceil((1.0 * success) / s);
            if (min > max) {
                ans[i++] = 0;
                continue;
            }
            ans[i++] = p - find(potions, (int)min);
        }

        return ans;
    }

    private int find(int[] potions, int s) {
        int l = 0, r = potions.length, m;
        while (l < r) {
            m = (r - l) / 2 + l;
            if (potions[m] < s) {
                l = m + 1;
            } else {
                r = m;
            }
        }
        return l;
    }
}